### PR TITLE
FileTarget: Use ManualResetEvent instead of Thread.Abort to stop appender-invalidator-thread

### LINKED ISSUE #1415: NLog calls Thread.Abort() to exit simple thread loop. It should not
Type: Bug
NLog. 4.3.3
Platform: all

Using Thread.Abort() is just bad practice. Especially for a simple processing loop. It would be better to use a ManualResetEvent, or its more efficient version ManualResetEventSlim

I have attached an example of how this would be implemented. I am not sure about the locking in the FileTarget as there does not seem to be any in this area, but I have added it just in case.

In general, I am not sure about the idea of spinning up a thread like this in the first place, but I do not know enough about the code to figure out a better way. In our application we have 4 log files, so there are 4 threads being spun up. These are looping every 200ms to do something very simple. Would a timer not be a better approach? Why a dedicated thread?

[FileTarget.txt](https://github.com/NLog/NLog/files/243333/FileTarget.txt)

### PR DESCRIPTION
**Pull request by bhaeussermann**
Fix #1415

### CODE REVIEW

**Comment by 304NotModified** [comment]
supersedes https://github.com/NLog/NLog/pull/1424?

**Comment by bhaeussermann** [comment]
> supersedes #1424?

Yes.

File: src/NLog/Targets/FileTarget.cs [hunk 1]
```diff
@@ -694,21 +694,23 @@ private void RefreshArchiveFilePatternToWatch()
...
+                                                break;
+                                            }
+
+                                            this.fileAppenderCache.InvalidateAppendersForInvalidFiles();
+                                        }
                                     }
                                     catch (Exception ex)
                                     {
-                                        InternalLogger.Warn(ex, "Exception in Thread.Sleep, most of the time not an issue.");
+                                        InternalLogger.Info(ex, "Exception in FileTarget appender-invalidator thread.");
```
Comment (304NotModified) [inline src/NLog/Targets/FileTarget.cs]: Maybe debug level is better?

File: src/NLog/Targets/FileTarget.cs [hunk 2]
```diff
@@ -719,17 +721,26 @@ private void RefreshArchiveFilePatternToWatch()
...
-                        this.appenderInvalidatorThread.Abort();
-                        this.appenderInvalidatorThread = null;
-                    }
+                    this.StopAppenderInvalidatorThread();
                 }
             }
 #endif
         }
 
+        private void StopAppenderInvalidatorThread()
```
Comment (304NotModified) [inline src/NLog/Targets/FileTarget.cs]: I think some docs are nice, as this is a pretty difficult case (to test at least)

**Comment by 304NotModified** [comment]
Some notes on docs and internal Logging. If I find time, I did this after merge.

**Comment by 304NotModified** [comment]
merged :)

**Pull request merged**